projects/npm-tools/packages/npm-scripts/src/utils/runWebpackAsBundler.js

```diff
@@ -3,82 +3,151 @@
  * SPDX-License-Identifier: BSD-3-Clause
  */
 
+/* eslint-disable @liferay/no-dynamic-require */
+
 const path = require('path');
+const resolve = require('resolve');
 const TerserPlugin = require('terser-webpack-plugin');
 
 const convertImportsToExternals = require('./convertImportsToExternals');
+const createTempFile = require('./createTempFile');
 const flattenPkgName = require('./flattenPkgName');
 const runWebpack = require('./runWebpack');
 
 /**
  * Runs webpack as a replacement of the bundler
  */
 module.exports = async function runWebpackAsBundler(config) {
-	const webpackConfig = getWebpackConfig(config);
+	const start = Date.now();
+
+	const webpackConfigs = getWebpackConfigs(config);
+
+	for (const webpackConfig of webpackConfigs) {
+		await runWebpack(webpackConfig, config.report);
+	}
+
+	const lapse = Math.floor((Date.now() - start) / 1000);
 
-	await runWebpack(webpackConfig);
+	/* eslint-disable-next-line no-console */
+	console.log(`ESM bundling took ${lapse}s`);
 };
 
-function getWebpackConfig(config) {
-	const {exports, imports} = config;
+function getEntryImportDescriptor(exportsItem) {
+	const pkgName = exportsItem.package;
+	const flatPkgName = flattenPkgName(pkgName);
+	let importPath = pkgName;
+
+	if (exportsItem.bridge) {
+		const exportObject = require(resolve.sync(pkgName, {basedir: '.'}));
+
+		const nonDefaultFields = Object.keys(exportObject)
+			.filter((field) => field !== 'default')
+			.map((field) => `	${field}`)
+			.join(',\n');
+
+		const bridgeSource = `
+import * as entryImport from '${pkgName}';
+
+const {
+${nonDefaultFields}
+} = entryImport;
+
+export {
+${nonDefaultFields}
+};
+`;
+		const {filePath} = createTempFile(`${flatPkgName}.js`, bridgeSource, {
+			autoDelete: false,
+		});
+
+		importPath = filePath;
+	}
 
 	return {
-		devtool: 'cheap-source-map',
-		entry: exports.reduce((entry, pkgName) => {
-			const flatPkgName = flattenPkgName(pkgName);
-
-			entry[`__liferay__/exports/${flatPkgName}`] = {
-				import: pkgName,
-			};
-
-			return entry;
-		}, {}),
-		experiments: {
-			outputModule: true,
-		},
-		externals: convertImportsToExternals(imports),
-		externalsType: 'module',
-		mode: 'development',
-		module: {
-			rules: [
-				{
-					exclude: /node_modules/,
-					test: /\.js$/,
-					use: {
-						loader: require.resolve('babel-loader'),
-						options: {
-							presets: [
-								require.resolve('@babel/preset-env'),
-								require.resolve('@babel/preset-react'),
-							],
+		flatPkgName,
+		importPath,
+		pkgName,
+	};
+}
+
+function getWebpackConfigs(config) {
+	const {exports, imports} = config;
+
+	return exports.reduce((webpackConfigs, exportsItem) => {
+		const {flatPkgName, importPath, pkgName} = getEntryImportDescriptor(
+			exportsItem
+		);
+
+		const externals = convertImportsToExternals(imports);
```

Comment:
Should this be hoisted outside of the reduce and inside of getWebpackConfigs? It doesn't depend on anything in this closure and the `delete` below made me think we want to mutate an object repeatedly and not just once.

Reply:
I could easily be wrong here, but it just stuck out to me

Reply:
> the delete below made me think we want to mutate an object repeatedly and not just once

Nope, it's the opposite. We want a fresh new object in each iteration. Then we delete just one property from that object.

But since I made the same reasoning as you and then had to undo the change locally to make it work, I'll rewrite it hoisting the `convertImportsToExternals`, then explicitly duplicating the object before `delete` :+1: .